### PR TITLE
o/snapstate: add core22 migration logic

### PR DESCRIPTION
**Pull request by MiguelPires**
Based on https://github.com/snapcore/snapd/pull/11413 so only ddeda073b7 (and later, if added) are relevant. Adds the migration logic for core22 and the AppArmor rules. The draft PR this was split off from is https://github.com/snapcore/snapd/pull/11383

### CODE REVIEW

File: overlord/snapstate/backend/copydata.go [hunk 1]
```diff
@@ -38,6 +39,14 @@ var (
 	mkdirAllChown = osutil.MkdirAllChown
 )
 
+func MockAllUsers(f func(options *dirs.SnapDirOptions) ([]*user.User, error)) func() {
+	old := allUsers
```
Comment (mardy) [inline overlord/snapstate/backend/copydata.go:43]: nitpick: add a call to `osutil.MustBeTestBinary`

File: overlord/snapstate/handlers.go [hunk 2]
```diff
@@ -1150,59 +1150,190 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
...
+	// none states that no action should be taken
+	none migration = "none"
+	// full migrates ~/snap to ~/.snap and the new home to ~/Snap
+	full migration = "full"
+	// hidden migrates ~/snap to ~/.snap
+	hidden migration = "hidden"
+	// home migrates the new home to ~/Snap
+	home migration = "home"
+	// revertHidden undoes the hidden migration (i.e., moves ~/.snap to ~/snap)
+	revertHidden migration = "revertHidden"
```
Comment (mardy) [inline overlord/snapstate/handlers.go:1197]: Will we have a `revertFull` as well? Just wondering, if instead of these strings we should have a bitfield with flags, since it seems to me that some of these actions can have some overlapping behaviour. (this is a question, not a request for change :-) )

**Review by mardy — APPROVED** [review]
LGTM, with a question.

File: overlord/snapstate/handlers.go (hunk 2 above)
Comment (MiguelPires) [reply overlord/snapstate/handlers.go:1197]: Yes, there will be a revertFull and even a revertHome so I agree that those would be well expressed as a bitfield. However, the remaining values can't be combined so we'd have to add more code to "forbid" that. One thing I maybe should've made a comment about is that these values are strings because it makes test failures easier to understand. They were iotas initially and the errors looked like "expected 0 got 5".

File: overlord/snapstate/backend/copydata.go (hunk 1 above)
Comment (MiguelPires) [reply overlord/snapstate/backend/copydata.go:43]: Will do. Thanks!

**Review by mardy — APPROVED** [review]
Thanks, LGTM!

File: interfaces/apparmor/template.go [hunk 3]
```diff
@@ -342,6 +342,9 @@ var templateCommon = `
   owner @{HOME}/.snap/data/@{SNAP_INSTANCE_NAME}/@{SNAP_REVISION}/** wl,
   owner @{HOME}/.snap/data/@{SNAP_INSTANCE_NAME}/common/**           wl,
 
+  owner @{HOME}/Snap/@{SNAP_INSTANCE_NAME}/                          r,
+  owner @{HOME}/Snap/@{SNAP_INSTANCE_NAME}/**                        mrkixwl,
```
Comment (pedronis) [inline interfaces/apparmor/template.go:346]: shouldn't this have the same permissions as @{HOME}/.snap/data/@{SNAP_INSTANCE_NAME}/@{SNAP_REVISION}/ ?

Comment (MiguelPires) [reply interfaces/apparmor/template.go:346]: Doesn't it? The files under the revisioned dir are affected by:
```
  owner @{HOME}/.snap/data/@{SNAP_INSTANCE_NAME}/**                  mrkix,
  owner @{HOME}/.snap/data/@{SNAP_INSTANCE_NAME}/@{SNAP_REVISION}/** wl,
```
Which looks equivalent to this unless I'm misinterpreting something. However, the directory itself needs to be corrected because the revisioned dir also has `mrkix` permissions and this one doesn't

**Review by pedronis — COMMENTED** [review]
didn't look closely at the tests yet but this overall looks good though the diff is quite hard to follow: one possibility about that would be to first refactor to the triggeredMigration and \*Migration helpers supporting only hidden, and then add ExposedHome functionality as its own separate commit/PR. Couple of remarks

File: overlord/snapstate/snapmgr.go [hunk 4]
```diff
@@ -115,6 +115,11 @@ type SnapSetup struct {
 	// the current change. A 'false' value doesn't mean the dir is hidden, just
 	// that it wasn't exposed in this change.
 	MigratedExposed bool `json:"migrated-exposed,omitempty"`
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply overlord/snapstate/snapmgr.go:117]: Yes, I think so, I noticed the same while working on the revert. How about `MigratedHidden`, `UndidHiddenMigration`, `MigratedToExposedHome` and `UndidHomeMigration`? The `Undid` prefix hopefully makes evident that they're reversions of the migrations

**Comment by MiguelPires** [comment]
@pedronis as mentioned, I re-organized the PR into a first commit that refactors the existing logic and a second commit that adds cases for full and home migration. I also applied your suggestions in those commits (as opposed to separate subsequent commits) since it already required force pushing and it'll make it easier to read the re-organized code

File: overlord/snapstate/handlers.go [hunk 5]
```diff
@@ -1150,59 +1150,178 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
...
+	full migration = "full"
+)
+
+func triggeredMigration(newBase string, opts *dirMigrationOptions) migration {
+	// we're refreshing to a core22 revision
+	if atLeastCore22(newBase) {
+		if opts.MigratedToHidden && !opts.MigratedToExposedHome {
+			// ~/.snap migration already happened so initialize ~/Snap only
+			return home
+		} else if !opts.MigratedToHidden {
```
Comment (stolowski) [inline overlord/snapstate/handlers.go:1201]: Can this just be `if !opts.MigratedToHidden` (dropping `else`)?

File: overlord/snapstate/handlers.go [hunk 6]
```diff
@@ -1150,59 +1150,178 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
...
+			//  nothing was migrated yet, so migrate to ~/.snap and ~/Snap
+			return full
+		}
+	} else {
+		if !opts.MigratedToHidden && opts.UseHidden {
+			// flag is set and not migrated yet
+			return hidden
+		} else if opts.MigratedToHidden && !opts.UseHidden {
```
Comment (stolowski) [inline overlord/snapstate/handlers.go:1209]: Same.

File: overlord/snapstate/handlers.go (hunk 5 above)
Comment (MiguelPires) [reply overlord/snapstate/handlers.go]: Sure. Dropped (and the other as well)

File: overlord/snapstate/handlers.go [hunk 7]
```diff
@@ -1150,59 +1150,178 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
...
-			return err
+	snapsup.MigratedHidden = true
+	if err = m.backend.HideSnapData(snapsup.InstanceName()); err != nil {
+		return undo, err
+	}
+
+	st.Lock()
+	err = SetTaskSnapSetup(t, snapsup)
+	st.Unlock()
+	if err != nil {
```
Comment (stolowski) [inline overlord/snapstate/handlers.go:1245]: Nitpick:
```suggestion
	st.Lock()
	defer st.Unlock()
	if err := SetTaskSnapSetup(t, snapsup); err != nil {
```

File: overlord/snapstate/handlers.go [hunk 8]
```diff
@@ -1150,59 +1150,178 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
...
-			return fmt.Errorf("cannot set migration status to undone: %w", err)
+	if err := m.backend.UndoHideSnapData(snapsup.InstanceName()); err != nil {
+		return undo, err
+	}
+
+	snapsup.UndidHiddenMigration = true
+	st.Lock()
+	err = SetTaskSnapSetup(t, snapsup)
+	st.Unlock()
+	if err != nil {
```
Comment (stolowski) [inline overlord/snapstate/handlers.go:1273]: Same, how about:
```suggestion
	st.Lock()
	defer st.Unlock()
	if err := SetTaskSnapSetup(t, snapsup); err != nil {
```

File: overlord/snapstate/handlers.go [hunk 9]
```diff
@@ -1150,59 +1150,178 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
...
 
-	return nil
+	snapsup.MigratedToExposedHome = true
+	if err := m.backend.InitExposedSnapHome(snapsup.InstanceName(), newRev); err != nil {
+		return undo, err
+	}
+	st.Lock()
+	err = SetTaskSnapSetup(t, snapsup)
+	st.Unlock()
+	if err != nil {
```
Comment (stolowski) [inline overlord/snapstate/handlers.go:1297]: And here.

File: overlord/snapstate/handlers.go [hunk 10]
```diff
@@ -3604,11 +3739,14 @@ type dirMigrationOptions struct {
 
 	// MigratedToHidden states whether the data has been migrated to the hidden dir
 	MigratedToHidden bool
+
+	// MigratedToExposedHome state whether the ~/Snap migration has been done.
```
Comment (stolowski) [inline overlord/snapstate/handlers.go:3747]:
```suggestion
	// MigratedToExposedHome states whether the ~/Snap migration has been done.
```

File: overlord/snapstate/snapmgr.go [hunk 11]
```diff
@@ -210,6 +215,10 @@ type SnapState struct {
 	// MigratedHidden is set if the user's snap dir has been migrated
 	// to ~/.snap/data.
 	MigratedHidden bool `json:"migrated-hidden,omitempty"`
+
+	// MigratedToExposedHome is set if ~/Snap was created and initialized. If set, ~/Snap
+	// should be used as the snap's HOME.
+	MigratedToExposedHome bool `json:"migrated-exposed-home"`
```
Comment (stolowski) [inline overlord/snapstate/snapmgr.go:221]: omitempty?

**Review by stolowski — COMMENTED** [review]
LGTM, a few nitpicks/suggestions.

**Review by stolowski — APPROVED** [review]

File: interfaces/apparmor/template.go (hunk 3 above)
Comment (alexmurray) [reply interfaces/apparmor/template.go:346]: These are not equivalent - note that the existing permissions for `~/.snap/data are`:

```
  # Experimental snap folder changes
  owner @{HOME}/.snap/data/@{SNAP_INSTANCE_NAME}/                    r,
  owner @{HOME}/.snap/data/@{SNAP_INSTANCE_NAME}/**                  mrkix,
  owner @{HOME}/.snap/data/@{SNAP_INSTANCE_NAME}/@{SNAP_REVISION}/** wl,
  owner @{HOME}/.snap/data/@{SNAP_INSTANCE_NAME}/common/**           wl,
```

Whereas you are now using:

```
  owner @{HOME}/Snap/@{SNAP_INSTANCE_NAME}/                          mrkix,
  owner @{HOME}/Snap/@{SNAP_INSTANCE_NAME}/**                        mrkixwl,
```

which allows more permissions in this case (since a snap now can `m`map,loc`k`,and e`x`ecute (with `i`nherit permissions) from the `@{HOME}/Snap/@{SNAP_INSTANCE_NAME}/` directory - which doesn't really make sense for a directory anyway, whereas before it could only `r` from `SNAP_INSTANCE_NAME`).

I think this should look more like:

```
  owner @{HOME}/Snap/@{SNAP_INSTANCE_NAME}/                  r,
  owner @{HOME}/Snap/@{SNAP_INSTANCE_NAME}/**                mrkixwl,
```

ie. read on just the directory but then the ability to create / execute / write arbitrary things within that path.

**Review by alexmurray — APPROVED** [review]
From a security point-of-view this seems ok other than correcting the apparmor permissions on the new `~/Snap/@{SNAP_INSTANCE_NAME}` directory.

Although it does seem to add a bunch of extra complexity without any real functional gain from what I can see which does then worry me a bit since with added complexity comes the risk of added latent security issues, not to mention the maintenance burden going forwards as well. But I will leave these larger issues to @mvo5 and @pedronis etc.

File: interfaces/apparmor/template.go (hunk 3 above)
Comment (MiguelPires) [reply interfaces/apparmor/template.go:346]: Ah, I see my mistake. Thank you, I've corrected the permissions in the latest commit

**Review by pedronis — COMMENTED** [review]
lgtm (post merge)
